Src/ILGPU/Runtime/Cuda/CudaPageLockScope.cs

```diff
@@ -32,25 +32,38 @@ public sealed class CudaPageLockScope<T> : PageLockScope<T>
             CudaAccelerator accelerator,
             IntPtr hostPtr,
             long numElements)
-            : base(accelerator)
+            : base(accelerator, numElements)
         {
             if (!accelerator.Device.SupportsMappingHostMemory)
             {
                 throw new NotSupportedException(
                     RuntimeErrorMessages.NotSupportedPageLock);
             }
             HostPtr = hostPtr;
-            Length = numElements;
 
+            bool supportsHostPointer = accelerator
+                .Device
+                .SupportsUsingHostPointerForRegisteredMemory;
+
+            // Setup internal memory registration flags.
             var flags = MemHostRegisterFlags.CU_MEMHOSTREGISTER_PORTABLE;
-            if (!accelerator.Device.SupportsUsingHostPointerForRegisteredMemory)
+            if (!supportsHostPointer)
                 flags |= MemHostRegisterFlags.CU_MEMHOSTREGISTER_DEVICEMAP;
-            CudaException.ThrowIfFailed(
-                CurrentAPI.MemHostRegister(
-                    hostPtr,
-                    new IntPtr(LengthInBytes),
-                    flags));
-            if (accelerator.Device.SupportsUsingHostPointerForRegisteredMemory)
+
+            // Perform the memory registration.
+            var cudaError = CurrentAPI.MemHostRegister(
+                hostPtr,
+                new IntPtr(LengthInBytes),
+                flags);
+
+            // Check whether we have already registered this memory pointer.
+            Locked = cudaError != CudaError.CUDA_ERROR_HOST_MEMORY_ALREADY_REGISTERED;
```

Comment:
ISSUE: When the memory address is already registered by another scope, there is no guarantee that the other scope will keep the registration longer than this scope.

Reply:
@MoFtZ Yes, you are right 🤓 I think we should revert these changes.

Reply:
Maybe we could introduce a reference-counting system?

Reply:
@MoFtZ I'm not sure it's worth introducing reference counting in this context.